File: AddTwoNumbers/AddTwoNumbers.py
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        # Create dummy head node
        dummy = ListNode(0)
        curr = dummy
        carry = 0
        
        # Process until both lists are empty and no carry remains
        while l1 or l2 or carry:
            # Get values
            x = l1.val if l1 else 0
            y = l2.val if l2 else 0
            
            # Calculate sum and carry
            total = x + y + carry
            carry = total // 10
            
            # Create new node
            curr.next = ListNode(total % 10)
            curr = curr.next
            
            # Move to next nodes if available
            if l1: l1 = l1.next
            if l2: l2 = l2.next
        
        return dummy.next

File: AddTwoNumbers/AddTwoNumbers.cpp

struct ListNode {
      int val;
      ListNode *next;
      ListNode() : val(0), next(nullptr) {}
      ListNode(int x) : val(x), next(nullptr) {}
      ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* dummy = new ListNode(0);
        ListNode* curr = dummy;
        int carry = 0;

        while( l1 || l2 || carry ){
            int value1 = ( l1 ? l1->val : 0 ) ;
            int value2 = ( l2 ? l2->val : 0 ) ;

            int total = value1 + value2 + carry;
            carry = total / 10;
            int digit = total % 10 ;

            curr->next = new ListNode(total % 10);
            curr = curr->next;

            if(l1) l1 = l1->next;
            if(l2) l2 = l2->next;
         }

         return dummy->next;
};

};